exchange/cache.go

```diff
@@ -0,0 +1,286 @@
+
+	if typedSvc, ok := svc.(map[string]ServiceDefinition); ok {
+		return typedSvc
+	}
+	return nil
+}
+
+// GetServicePolicyFromCache returns the service policy from the exchange cache if it is present, or nil if it is not
+func GetServicePolicyFromCache(svcOrg string, svcId string, svcArch string, svcVersion string) *ExchangePolicy {
+	svcPol := GetResourceFromCache(ServicePolicyCacheMapKey(svcOrg, svcId, svcArch, svcVersion), SVC_POL_TYPE_CACHE, 0)
+
+	if typedSvcPol, ok := svcPol.(ExchangePolicy); ok {
+		return &typedSvcPol
+	}
+	return nil
+}
+
+// GetExchangeVersionFromCache returns the version of the exchange from the exchange cache if it is present or an emty string otherwise
+func GetExchangeVersionFromCache(exchangeURL string) string {
+	exchVers := GetResourceFromCache(exchangeURL, EXCH_VERS_TYPE_CACHE, CACHE_TIMEOUT_S)
+
+	if typedExchVers, ok := exchVers.(string); ok {
+		return typedExchVers
+	}
+	return ""
+}
+
+// GetResourceFromCache will return the requested resource from the specified type exchange cache or nil if it is not present
+func GetResourceFromCache(resourceKey string, resourceType string, expirationS uint64) interface{} {
+	glog.V(5).Infof("Get from exchange cache %s/%s", resourceType, resourceKey)
+
+	if ExchangeResourceCache == nil || ExchangeResourceCache.allResources == nil {
+		return nil
+	}
+
+	ExchangeResourceCache.Lock.Lock()
+	defer ExchangeResourceCache.Lock.Unlock()
+
+	resourceCache, ok := ExchangeResourceCache.allResources[resourceType]
+	if !ok {
+		return nil
+	}
+	entry := resourceCache.Get(resourceKey)
+	if entry == nil {
+		return nil
+	}
+	typedEntry, ok := entry.(CacheEntry)
+	if !ok {
+		glog.Errorf("Error: object returned from cache not of expected type.")
+	}
+	expired := uint64(time.Now().Unix())-typedEntry.LastUpdated > expirationS
+	if expirationS > 0 && expired {
+		return nil
+	}
+	return typedEntry.Resource
+}
+
+// UpdateCache will replace or create the provided resource in the given resource type cache
+func UpdateCache(resourceKey string, resourceType string, updatedResource interface{}) {
+	glog.V(5).Infof("Update exchange cache %s/%s", resourceType, resourceKey)
+
+	if ExchangeResourceCache == nil {
+		newExchangeResourceCache := NewResourceCache()
+		ExchangeResourceCache = &newExchangeResourceCache
+	}
+
+	ExchangeResourceCache.Lock.Lock()
+	defer ExchangeResourceCache.Lock.Unlock()
+
+	resourceCache, ok := ExchangeResourceCache.allResources[resourceType]
+	if !ok {
+		ExchangeResourceCache.allResources[resourceType] = cache.NewSimpleMapCache()
+		resourceCache = ExchangeResourceCache.allResources[resourceType]
+	}
+	recordHash, err := hashResource(updatedResource)
+	if err != nil {
+		glog.Errorf("Failed to hash resource for cache. Error was : %v", err)
+		recordHash = []byte{}
+	}
+	existingRecord := resourceCache.Get(resourceKey)
+	existingRecordTyped := CacheEntry{}
+	if existingRecord != nil {
+		existingRecordTyped = existingRecord.(CacheEntry)
```

Comment:
We might need to protect this cast with a check to prevent a panic if the cast fails.

Reply:
Fixed